www/js/lib/zimArchive.js

```diff
@@ -190,12 +232,15 @@ define(['zimfile', 'zimDirEntry', 'util', 'utf8'],
         }, true).then(function(firstIndex) {
             var dirEntries = [];
             var addDirEntries = function(index) {
-                if (index >= firstIndex + resultSize || index >= that._file.articleCount)
+                if (search.state === 'cancelled' || index >= firstIndex + resultSize || index >= that._file.articleCount)
                     return dirEntries;
                 return that._file.dirEntryByTitleIndex(index).then(function(dirEntry) {
                     var title = dirEntry.getTitleOrUrl();
-                    if (!title.indexOf(prefix) && dirEntry.namespace === "A")
+                    if (~title.indexOf(prefix) && dirEntry.namespace === "A") {
                         dirEntries.push(dirEntry);
+                        // Report interim result
+                        callback(dirEntry, true);
```

Comment:
We should call back `[dirEntry]` here, i.e. dirEntry should be in an array, not a bare object, even though it works with concat.